File: java/src/main/java/dev/sanmer/sac/io/Sac.java
package dev.sanmer.sac.io;

import java.io.Closeable;
import java.io.File;

public class Sac implements Closeable {
    private final long ptr;

    public Sac(long ptr) {
        this.ptr = ptr;
    }

    public SacHeader getHeader() {
        return getHeader(ptr);
    }

    public void setHeader(SacHeader value) {
        setHeader(ptr, value);
    }

    public float[] getFirst() {
        return getFirst(ptr);
    }

    public void setFirst(float[] value) {
        setFirst(ptr, value);
    }

    public float[] getSecond() {
        return getSecond(ptr);
    }

    public void setSecond(float[] value) {
        setSecond(ptr, value);
    }

    public void writeHeader() {
        writeHeader(ptr);
    }

    public void setEndian(Endian endian) {
        setEndian(ptr, endian.ordinal());
    }

    public void write() {
        write(ptr);
    }

    public void writeTo(File file) {
        writeTo(ptr, file.getAbsolutePath());
    }

    @Override
    public void close() {
        drop(ptr);
    }

    static {
        Library.load();
    }

    private static native long readHeader(String path, int endian);

    private static native long read(String path, int endian);

    private static native long empty(String path, int endian);

    private static native void writeHeader(long ptr);

    private static native void write(long ptr);

    private static native void writeTo(long ptr, String path);

    private static native SacHeader getHeader(long ptr);

    private static native void setHeader(long ptr, SacHeader h);

    private static native float[] getFirst(long ptr);

    private static native void setFirst(long ptr, float[] x);

    private static native float[] getSecond(long ptr);

    private static native void setSecond(long ptr, float[] y);

    private static native void setEndian(long ptr, int endian);

    private static native void drop(long ptr);

    public static Sac readHeader(File file, Endian endian) {
        long ptr = readHeader(file.getAbsolutePath(), endian.ordinal());
        return new Sac(ptr);
    }

    public static Sac read(File file, Endian endian) {
        long ptr = read(file.getAbsolutePath(), endian.ordinal());
        return new Sac(ptr);
    }

    public static Sac empty(File file, Endian endian) {
        long ptr = empty(file.getAbsolutePath(), endian.ordinal());
        return new Sac(ptr);
    }

}

File: kotlin/src/main/kotlin/dev/sanmer/sac/io/Sac.kt
package dev.sanmer.sac.io

import java.io.Closeable
import java.io.File

class Sac(
    private val ptr: Long
) : Closeable {
    var h: SacHeader
        get() = getHeader(ptr)
        set(value) = setHeader(ptr, value)

    var x: FloatArray
        get() = getX(ptr)
        set(value) = setX(ptr, value)

    var y: FloatArray
        get() = getY(ptr)
        set(value) = setY(ptr, value)

    fun writeHeader() = writeHeader(ptr)
    fun setEndian(endian: Endian) = setEndian(ptr, endian.ordinal)
    fun write() = write(ptr)
    fun writeTo(file: File) = writeTo(ptr, file.absolutePath)
    override fun close() = drop(ptr)

    companion object {
        init {
            SacLibrary.load()
        }

        @JvmStatic
        private external fun readHeader(path: String, endian: Int): Long

        @JvmStatic
        private external fun read(path: String, endian: Int): Long

        @JvmStatic
        private external fun writeHeader(ptr: Long)

        @JvmStatic
        private external fun write(ptr: Long)

        @JvmStatic
        private external fun writeTo(ptr: Long, path: String)

        @JvmStatic
        private external fun getHeader(ptr: Long): SacHeader

        @JvmStatic
        private external fun setHeader(ptr: Long, h: SacHeader)

        @JvmStatic
        private external fun getX(ptr: Long): FloatArray

        @JvmStatic
        private external fun setX(ptr: Long, x: FloatArray)

        @JvmStatic
        private external fun getY(ptr: Long): FloatArray

        @JvmStatic
        private external fun setY(ptr: Long, y: FloatArray)

        @JvmStatic
        private external fun setEndian(ptr: Long, endian: Int)

        @JvmStatic
        private external fun drop(ptr: Long)

        fun readHeader(file: File, endian: Endian): Sac {
            val ptr = readHeader(
                path = file.absolutePath,
                endian = endian.ordinal
            )

            return Sac(ptr)
        }

        fun read(file: File, endian: Endian): Sac {
            val ptr = read(
                path = file.absolutePath,
                endian = endian.ordinal
            )

            return Sac(ptr)
        }
    }
}